dissect/volume/ddf/ddf.py

```diff
@@ -0,0 +1,369 @@
+
+from dissect.cstruct.types import Instance, Structure
+from dissect.util import ts
+
+from dissect.volume.ddf.c_ddf import DEFAULT_SECTOR_SIZE, c_ddf
+from dissect.volume.exceptions import DDFError
+from dissect.volume.raid.raid import RAID, Configuration, PhysicalDisk, VirtualDisk
+from dissect.volume.raid.stream import Layout, Level
+
+DECADE = 3600 * 24 * (365 * 10 + 2)
+
+
+class DDF(RAID):
+    def __init__(self, fh: list[Union[BinaryIO, DDFPhysicalDisk]], sector_size: int = DEFAULT_SECTOR_SIZE):
+        fhs = [fh] if not isinstance(fh, list) else fh
+        self.disks = [DDFPhysicalDisk(f, sector_size) if not isinstance(f, DDFPhysicalDisk) else f for f in fhs]
+        self.sector_size = sector_size
+
+        config_map = {}
+        for pd in self.disks:
+            config_map.setdefault(pd.anchor.DDF_Header_GUID, []).append(pd)
+
+        super().__init__([DDFConfiguration(disks) for disks in config_map.values()])
+
+
+class DDFConfiguration(Configuration):
+    def __init__(self, fh: list[Union[BinaryIO, PhysicalDisk]], sector_size: int = DEFAULT_SECTOR_SIZE):
+        fhs = [fh] if not isinstance(fh, list) else fh
+        self.disks = [DDFPhysicalDisk(f, sector_size) if not isinstance(f, DDFPhysicalDisk) else f for f in fhs]
+        self.sector_size = sector_size
+
+        pd_map: dict[int, DDFPhysicalDisk] = {}
+        vde_map: dict[bytes, VirtualDiskRecord] = {}
+        vdcr_map: dict[bytes, VirtualDiskConfigurationRecord] = {}
+        vdcr_uniq: dict[tuple[bytes, int], VirtualDiskConfigurationRecord] = {}
+
+        for pd in self.disks:
+            pd_map[pd.reference] = pd
+            vde_map.update({vde.guid: vde for vde in pd.virtual_disk_records})
+            vdcr_map.update({vdcr.guid: vdcr for vdcr in pd.virtual_disk_configuration_records})
+            vdcr_uniq.update(
+                {(vdcr.guid, vdcr.secondary_element_seq): vdcr for vdcr in pd.virtual_disk_configuration_records}
+            )
+
+        vd_map: dict = {}
+        for vdcr in vdcr_uniq.values():
+            vd_pd_map = vd_map.setdefault(vdcr.guid, {})
+
+            count = vdcr.primary_element_count
+            sec = 0 if vdcr.secondary_element_count == 1 else vdcr.secondary_element_seq
+
+            i = 0
+            for starting_block, pds in zip(vdcr.starting_block, vdcr.physical_disk_sequence):
+                if matched_pd := pd_map.get(pds):
+                    vd_pd_map[sec * count + i] = (starting_block, matched_pd)
+
+                if pds != 0xFFFFFFFF:
+                    i += 1
+
+        virtual_disks = [DDFVirtualDisk(vdcr_map[guid], vde_map[guid], vd_map[guid]) for guid in vd_map.keys()]
+        super().__init__(self.disks, virtual_disks)
+
+
+class DDFVirtualDisk(VirtualDisk):
+    def __init__(
+        self,
+        vdcr: VirtualDiskConfigurationRecord,
+        vdr: VirtualDiskRecord,
+        disks: dict[int, tuple[int, DDFPhysicalDisk]],
+    ):
+        self.vdcr = vdcr
+        self.vdr = vdr
```

Comment:
Since these are public properties, it might be better to give them more descriptive names?

Reply:
I've added properties with the long name variants, is that alright?